Baremetal v1 API: Nodes

stbenjam opened the pull request:
For #1429 

First PR to add support for Ironic to gophercloud.  This contains the basics for node support. Node management API will be done separately, along with the other API objects as listed in #1429.

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:

General:
- [API Controller](https://github.com/openstack/ironic/blob/master/ironic/api/controllers/v1/node.py)

Create:
- [API Reference](https://developer.openstack.org/api-ref/baremetal/?expanded=create-node-detail#create-node)
- [`Node` class definition code](https://github.com/openstack/ironic/blob/master/ironic/api/controllers/v1/node.py#L885-L1080)
- [Create API endpoint code](https://github.com/openstack/ironic/blob/master/ironic/api/controllers/v1/node.py#L1958-L2023)

List:
- [API Reference](https://developer.openstack.org/api-ref/baremetal/?expanded=list-nodes-detail#create-node)
- [List Filters](https://github.com/openstack/ironic/blob/master/ironic/api/controllers/v1/node.py#L1627-L1649)
- [ProvisionState filter](https://github.com/openstack/ironic/blob/master/ironic/common/states.py#L66-L244)

jtopjian commented on openstack/baremetal/v1/nodes/doc.go:
```diff
@@ -0,0 +1,5 @@
+Package nodes provides information and interaction with the nodes API
+resource in the OpenStack Bare Metal service.
+*/
+package nodes
```
This file needs to have examples added to it of each action you are adding in this PR.

jtopjian commented on openstack/baremetal/v1/nodes/errors.go:
```diff
@@ -0,0 +1 @@
+package nodes
```
This file doesn't seem to be used? So let's remove it.

jtopjian commented on openstack/baremetal/v1/nodes/requests.go:
```diff
@@ -0,0 +1,232 @@
+// CreateOpts specifies node creation parameters.
+type CreateOpts struct {
+	// The boot interface for a Node, e.g. “pxe”.
+	BootInterface string `json:"boot_interface,omitempty"`
```
Are there a finite amount of valid options? If so, then create a custom type for this:

```go
type BootInterface string

const (
  PXEBoot BootInterface = "pxe"
  FooBoot BootInterface = "foo"
)
```

This goes for any other field in CreateOpts with finite types. In the `results.go` file, do not use the custom type - just use `string`.

stbenjam replied:
I believe these are not finite options, because of Ironic drivers for different hardware platforms.  A driver can add additional types of boot interfaces (and other things).  ProvisionState in ListOpts does seem to be at least one that has fixed values and I have added a type for those.

jtopjian commented on openstack/baremetal/v1/nodes/requests.go:
```diff
@@ -0,0 +1,232 @@
+	StorageInterface string `json:"storage_interface,omitempty"`
+
+	// The UUID for the resource.
+	Uuid string `json:"uuid,omitempty"`
```
`s/Uuid/UUID/`

jtopjian commented on openstack/baremetal/v1/nodes/urls.go:
```diff
@@ -0,0 +1,59 @@
+package nodes
```
Only include the URLs being used in this PR.

stbenjam replied:
Removed unused URLs

jtopjian commented on openstack/client.go:
```diff
@@ -307,6 +307,12 @@ func initClientOpts(client *gophercloud.ProviderClient, eo gophercloud.EndpointO
 
+// NewBareMetal creates a ServiceClient that may be used with the v1
+// bare metal package.
+func NewBareMetal(client *gophercloud.ProviderClient, eo gophercloud.EndpointOpts) (*gophercloud.ServiceClient, error) {
```
`NewBareMetalVn` where n is the API version

stbenjam commented on openstack/baremetal/v1/nodes/requests.go:
```diff
@@ -0,0 +1,274 @@
+
+	// All the metadata required by the driver to manage this Node. List of fields varies between drivers, and can
+	// be retrieved from the /v1/drivers/<DRIVER_NAME>/properties resource.
+	DriverInfo map[string]string `json:"driver_info,omitempty"`
```
One thing I'm concerned about here is that Ironic considers this field to just be arbitrary JSON.  There's a few fields like that.  They are typically just `map[string]string` but I haven't dug into all the known drivers yet.  I wonder if gophercloud shouldn't know more about the drivers and type specific structs for e.g. HP iLO driver info.

jtopjian replied:
We ran into that with Senlin/Clustering. For times when it was possible to define a strict structure, we implemented it. For other times, the field type ended up being either `interface{}` or `map[string]interface{}` - totally acceptable if that happens.

stbenjam commented:
Thanks for taking the time to review my PR. This should be in better shape for review. I've added some basic acceptance testing - is there Ironic available in the environment there?

I've also added links to the relevant Ironic code for create and list.  I do think the API would accept some additional fields in `CreateOpts`, like UUID but that's not specified in the API ref and I think not intended for end user use.  The Ironic API reference does seem to be very well maintained. I've included links to it, as well.

stbenjam commented:
Also what should I do about the coveralls failures? I think things like my addition to the service client is causing the decrease in coverage: https://coveralls.io/builds/21468077/source?filename=service_client.go

I could add some unit tests to cover the SC.

jtopjian commented:
> Also what should I do about the coveralls failures?

You can safely ignore coveralls :)

> is there Ironic available in the environment there?

No, OpenLab doesn't support Ironic at this time. Including acceptance tests and verifying they've worked in your own environment is good enough.

Thank you for making all of the requested changes! I will have time to review this in detail today or tomorrow.

jtopjian commented on openstack/baremetal/v1/nodes/requests.go:
```diff
@@ -0,0 +1,274 @@
+	Properties map[string]interface{} `json:"properties,omitempty"`
+
+	// Interface used for configuring RAID on this node, e.g. “no-raid”.
+	RaidInterface string `json:"raid_interface,omitempty"`
```
nit: should this be `RAIDInterface`?

jtopjian commented on openstack/baremetal/v1/nodes/results.go:
```diff
@@ -0,0 +1,222 @@
+	InstanceInfo map[string]interface{} `json:"instance_info"`
+
+	// ID of the Nova instance associated with this Node.
+	InstanceUuid string `json:"instance_uuid"`
```
`InstanceUUID`

jtopjian commented on openstack/baremetal/v1/nodes/results.go:
```diff
@@ -0,0 +1,222 @@
+	InstanceUuid string `json:"instance_uuid"`
+
+	// ID of the chassis associated with this Node. May be empty or None.
+	ChassisUuid string `json:"chassis_uuid"`
```
`ChassisUUID`

jtopjian commented on openstack/baremetal/v1/nodes/results.go:
```diff
@@ -0,0 +1,222 @@
+	ConsoleEnabled bool `json:"console_enabled"`
+
+	// The current RAID configuration of the node. Introduced with the cleaning feature.
+	RaidConfig map[string]interface{} `json:"raid_config"`
```
`RAIDConfig`?

jtopjian commented on openstack/baremetal/v1/nodes/results.go:
```diff
@@ -0,0 +1,222 @@
+
+	// The requested RAID configuration of the node, which will be applied when the Node next transitions
+	// through the CLEANING state. Introduced with the cleaning feature.
+	TargetRaidConfig map[string]interface{} `json:"target_raid_config"`
```
`TargetRAIDConfig`?

jtopjian commented on openstack/baremetal/v1/nodes/results.go:
```diff
@@ -0,0 +1,222 @@
+	PowerInterface string `json:"power_interface"`
+
+	// Used for configuring RAID on this node, e.g. “no-raid”.
+	RaidInterface string `json:"raid_interface"`
```
`RAIDInterface`?

stbenjam commented:
Great, thank you so much for the reviews! I've updated the acronyms to be uppercase - I'll be more careful of that on future PR's.

dtantsur commented:
Hi! Is there any code to provide/detect microversions here? This patch adds a lot of features that are only available with a microversion.

jtopjian commented:
@dtantsur That's a good point to bring up.

Microversions are really, really hard to work with in Go. There are a handful of areas where they are already being implemented. Usually this involves placing the responsibility of the developer using Gophercloud to make sure microversion-added features are used correctly (ex: if a field is set, make sure the client has the relevant microversion set).

We have some rough guidelines on how to handle microversions here: https://github.com/gophercloud/gophercloud/blob/master/docs/MICROVERSIONS.md

Let me know if this changes anything with how this PR should be structured. I'll hold off on merging for now.

stbenjam commented:
Not sure what you're looking for exactly, but yes this PR provides the ability to set the Ironic microversion by setting it on the service clients, that's shown in the examples in doc.go for baremetal.  All the various fields are marked omitempty if not specified.  That's basically the sum of microversion support in gophercloud at the moment until the proposals in https://github.com/gophercloud/gophercloud/issues/441 become a reality.

jtopjian commented:
> until the proposals in #441 become a reality.

Just a side note, the doc I posted is the result of those discussions. It was merged a few days ago and I forgot to close that issue.

But thank you so much for reading through all of that history :)

jtopjian commented:
@stbenjam To confirm, I think you're saying that you've kept microversions in mind as best as possible? If so, This all LGTM and I'll merge it.

stbenjam commented:
@jtopjian I think so.  I've tried my changes on multiple Ironic versions and they work, even if they don't know about certain fields (e.g. `owner` is new). Users still have to know to not set unsupported fields in CreateOpts, and also realize they won't get populated on older microversions.  However, I believe that's in line with what I see from the other services here.

@dtantsur would know better than I if there needs to be any other special handling other than that.

jtopjian commented:
OK, sounds good. I'll leave this open for a little.

dtantsur commented on openstack/baremetal/v1/nodes/requests.go:
```diff
@@ -0,0 +1,274 @@
+}
+
+// List makes a request against the API to list nodes accessible to you.
+func List(client *gophercloud.ServiceClient, opts ListOptsBuilder) pagination.Pager {
```
Something I don't quite get here is whether you handle detailed/non-detailed queries. I.e. in Ironic (cargo culted from Nova) we have /v1/nodes returning a list of summaries, /v1/nodes/detail returning a list of full nodes. In some API version we also introduced /v1/nodes?detail=True. Is it handled here?

stbenjam replied:
To get the list via `/v1/nodes?detail=True`:

```go
nodes.List(client, nodes.ListOpts{
    Detail: true,
})
```
However, the [API docs say `/v1/nodes/detail` is deprecated](https://developer.openstack.org/api-ref/baremetal/?expanded=list-nodes-detailed-detail) so I did not implement it - should I?

dtantsur replied:
It depends on which versions of Ironic you target. detail=True is only available since Rocky, which is quite recent. I guess to support older versions we still need /v1/nodes/detail, maybe with fallback to /v1/nodes?detail=True if microversion is large enough.

jtopjian replied:
This is most likely relevant to this discussion: https://github.com/gophercloud/gophercloud/issues/1440

stbenjam replied:
Sort of, except `/v1/nodes/detail` is deprecated, the modern way works as in the example code above, because the correct URL is `/v1/nodes?detail=True`. 

What to do about older Ironic is getting into microversions discussion that I would rather not try to solve.

I believe the correct way to do this is have my code be aware of the microversion, and do the right thing for older Ironics if Detail is set in ListOpts, by hitting the different URL.  That requires comparison of microversion, which I think is probably would need pulling in something like https://github.com/hashicorp/go-version to correctly compare versions.

dtantsur replied:
If you'd prefer to avoid diving into microversions here and now, you need support for the deprecated endpoint, since it works for all existing microversions.

stbenjam replied:
There is no code in gophercloud today to handling specific microversion cases which require comparison, and because this library supports many other things other than just Ironic I think it's a better thing to happen in a dedicated PR as I may not consider all the issues here for everyone. 

For the detail endpoint specifically, I don't like how it looks with two different ways to do things with no hints to the user which they should use. One way to avoid that is I *could* remove the Detail option from ListOpts, and have everyone go through the ListDetail function for a detailed listing.  Once the above concerns about microversions are addressed, add the microversion support later to hit the new or old URLs, but for now hit `/v1/nodes/detail` for everyone.  What I want to avoid is having two ways to show listing with details.  Does that make sense?

dtantsur replied:
Yes, I agree.

jtopjian replied:
> I believe the correct way to do this is have my code be aware of the microversion,

Your application code should be aware of microversions.

> For the detail endpoint specifically, I don't like how it looks with two different ways to do things with no hints to the user which they should use.

The user should interact with an application which interacts with Gophercloud. The application should contain the logic which determines how to use Gophercloud.

Think of Gophercloud as one step above a generic REST library.

stbenjam replied:
Sorry - user in this context was developer interacting with gophercloud directly. I've pushed some changes that I think address the concerns here.

jtopjian commented on openstack/baremetal/v1/nodes/requests.go:
```diff
@@ -0,0 +1,274 @@
+
+	// All the metadata required by the driver to manage this Node. List of fields varies between drivers, and can
+	// be retrieved from the /v1/drivers/<DRIVER_NAME>/properties resource.
+	DriverInfo map[string]interface{} `json:"driver_info,omitempty"`
```
Just to clarify: `omitempty` will cause the `driver_info` field to be omitted from the request body if it has no value.

dtantsur replied:
I was going to say that this field always has a value, then remembered that we can do things like `/v1/nodes?fields=name,driver`. Which btw will break any assumptions about required fields.

stbenjam replied:
Ok but this is CreateOpts, fields isn't relevant here.  Your example is a listing, whatever API sends back is populated, required/omitempty has no bearing there.

dtantsur replied:
Right, right. I've confused everything :)

jtopjian commented on openstack/baremetal/v1/nodes/results.go:
```diff
@@ -0,0 +1,222 @@
+	TargetPowerState string `json:"target_power_state"`
+
+	// Current provisioning state of this Node.
+	ProvisionState string `json:"provision_state"`
```
Right. Custom types (ie `ProvisionState`) are only used in requests to limit the valid input when it's possible to do so. For output/results, standard types are used to account for missed, new, or custom values.

stbenjam commented on openstack/baremetal/v1/nodes/doc.go:
```diff
@@ -0,0 +1,65 @@
+	// Example to Create Node
+	createNode, err := nodes.Create(client, nodes.CreateOpts{
+		Driver:        "ipmi",
+		BootInterface: PXEBoot,
```
There was a discussion about this on an earlier version of the PR - https://github.com/gophercloud/gophercloud/pull/1431#discussion_r253270627, which I decided against because I think new ones could be easily introduced by drivers, right?

In the end, I basically came down to driver supplied values to be as flexible as possible, but Ironic core like provisioning states could be constants - does that make sense?

stbenjam replied:
(I'll update the docs unless you think PXEBoot/DiskBoot/etc would be good to have)

dtantsur replied:
I think we should allow arbitrary things, so let's fix the docs.

stbenjam commented on openstack/baremetal/v1/nodes/requests.go:
```diff
@@ -0,0 +1,274 @@
+	RescueFail   ProvisionState = "rescue failed"
+	Rescuing     ProvisionState = "rescuing"
+	UnrescueFail ProvisionState = "unrescue failed"
+)
```
This was similar to the PXEBoot discussion above.

The actual `Node` struct uses string for this value so everything would be deserialized correctly if something new is introduced, but ListOpts takes known values, and these could also be used to compare, e.g. `node.ProvisionState == Rescuing` would work. 

I don't think anything would break, but we'd need to declare a new ProvisionState for it to work in list filters

stbenjam commented on openstack/baremetal/v1/nodes/requests.go:
```diff
@@ -0,0 +1,274 @@
+type UpdateOperation struct {
+	Op    UpdateOp `json:"op,required"`
+	Path  string   `json:"path,required"`
+	Value string   `json:"value,required"`
```
Oh good catch, thanks

stbenjam commented on openstack/baremetal/v1/nodes/results.go:
```diff
@@ -0,0 +1,222 @@
+
+	// Reason the node is marked as protected.
+	ProtectedReason string `json:"protected_reason"`
+}
```
Fixed

stbenjam commented:
@dtantsur Thanks for looking, responses above.  I also pushed fixes for the docs, missing owner, and `PATCH`.

stbenjam commented:
@dtantsur Could you look again please? I would like this to get in soon so I can open some additional pull requests for what we need in metalkube.

dtantsur commented on openstack/baremetal/v1/nodes/doc.go:
```diff
@@ -4,10 +4,24 @@ package nodes
 	nodes.List(client, nodes.ListOpts{
-		Detail: true,
 		ProvisionState: Deploying,
+		Fields: []string{"name"},
```
Is this going to work? I think you treat UUID as required.

stbenjam replied:
What do you mean? ListOpts has no required fields.

dtantsur replied:
I'm just asking if your code is ready to handle nodes with UUID missing. I thought no, but now I've skimmed through the code and it seems okay to me.

stbenjam replied:
The `Node` struct is pretty happy to deserialize everything I've seen the Ironic API throw at it.

dtantsur commented on openstack/baremetal/v1/nodes/requests.go:
```diff
@@ -9,37 +11,38 @@ import (
-	RescueFail   ProvisionState = "rescue failed"
-	Rescuing     ProvisionState = "rescuing"
-	UnrescueFail ProvisionState = "unrescue failed"
+	Verifying                   = "verifying"
```
I'm not sure I really get why this was changed in this commit..

stbenjam replied:
Unrelated style change, this is all going to get squashed anyway...

stbenjam commented:
I think this should be ready to go, let me know if you'd like me to make any additional changes. Thanks!

stbenjam commented:
Thanks!!!
